drivers/input/input_cap12xx.c

```diff
@@ -0,0 +1,229 @@
+
+	cap12xx_process(data->dev);
+}
+
+#ifdef CONFIG_INPUT_CAP12XX_POLL
+static void cap12xx_timer_handler(struct k_timer *timer)
+{
+	struct cap12xx_data *data = CONTAINER_OF(timer, struct cap12xx_data, timer);
+
+	k_work_submit(&data->work);
+}
+#else /* INTERRUPT MODE */
+static void cap12xx_isr_handler(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct cap12xx_data *data = CONTAINER_OF(cb, struct cap12xx_data, int_gpio_cb);
+
+	k_work_submit(&data->work);
+}
+#endif
+
+static int cap12xx_init(const struct device *dev)
+{
+	const struct cap12xx_config *config = dev->config;
+	struct cap12xx_data *data = dev->data;
+	int r;
+
+	if (!device_is_ready(config->i2c.bus)) {
+		LOG_ERR("I2C controller device not ready");
+		return -ENODEV;
+	}
+
+	data->dev = dev;
+
+	k_work_init(&data->work, cap12xx_work_handler);
+
+#ifdef CONFIG_INPUT_CAP12XX_POLL
+	LOG_DBG("cap12xx driver in polling mode");
+	k_timer_init(&data->timer, cap12xx_timer_handler, NULL);
+	r = cap12xx_enable_interrupt(&config->i2c, true);
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt");
+		return r;
+	}
+	k_timer_start(&data->timer, K_MSEC(CONFIG_INPUT_CAP12XX_PERIOD),
+		      K_MSEC(CONFIG_INPUT_CAP12XX_PERIOD));
+#else /* INTERRUPT MODE */
+	LOG_DBG("cap12xx driver in interrupt mode");
+	if (!gpio_is_ready_dt(&config->int_gpio)) {
+		LOG_ERR("Interrupt GPIO controller device not ready (missing device tree node?)");
+		return -ENODEV;
+	}
+
+	r = gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT);
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt GPIO pin");
+		return r;
+	}
+
+	r = gpio_pin_interrupt_configure_dt(&config->int_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt GPIO interrupt");
+		return r;
+	}
+
+	gpio_init_callback(&data->int_gpio_cb, cap12xx_isr_handler, BIT(config->int_gpio.pin));
+
+	r = gpio_add_callback(config->int_gpio.port, &data->int_gpio_cb);
+	if (r < 0) {
+		LOG_ERR("Could not set gpio callback");
+		return r;
+	}
+
+	r = cap12xx_clear_interrupt(&config->i2c);
+	if (r < 0) {
+		LOG_ERR("Could not clear interrupt");
+		return r;
+	}
+	r = cap12xx_enable_interrupt(&config->i2c, true);
+	if (r < 0) {
+		LOG_ERR("Could not configure interrupt");
+		return r;
+	}
+	if (config->repeat) {
+		r = i2c_reg_write_byte_dt(&config->i2c, REG_REPEAT_ENABLE, REPEAT_ENABLE);
+		if (r < 0) {
+			LOG_ERR("Could not disable repeated interrupts");
+			return r;
+		}
+		LOG_DBG("cap12xx enabled repeated interrupts");
+	} else {
+		r = i2c_reg_write_byte_dt(&config->i2c, REG_REPEAT_ENABLE, REPEAT_DISABLE);
+		if (r < 0) {
+			LOG_ERR("Could not enable repeated interrupts");
+			return r;
+		}
+		LOG_DBG("cap12xx disabled repeated interrupts");
+	}
+#endif
+	return 0;
+}
+
+#define CAP12XX_INIT(index)                                                                        \
+	BUILD_ASSERT(DT_INST_PROP_LEN(index, input_codes) == DT_INST_PROP(index, channels));       \
```

Comment:
The BUILD_ASSERT macro errors are unreadable (at least hard to decode) when no message is added, so I have this:
BUILD_ASSERT(DT_INST_NODE_HAS_PROP(index, int_gpios), "cap12xx: device tree node needs int-gpios property");
It seems to work, needs more testing later.

Adding a dependency to !CONFIG_INPUT_CAP12XX_POLL is hard, because that is not defined when not selected.
I just wonder if I should remove the Kconfig option altogether and move to a device tree only config?
Then I would need a BUILD_ASSERT(DT_INST_NODE_HAS_PROP(index, int_gpios) "and not" DT_INST_NODE_HAS_PROP(index, poll_interval),"cap12xx: use either interrupt or polling mode")
(how do I do "and not" in C-preprocessor statements?)

If I remove the Kconfig polling option, should I define a "CAP12XX_POLL" depending on DT_INST_NODE_HAS_PROP(index, poll_interval) at the top?

With more macro magic, I could remove the "channels" property of the device tree and use the property length of the "input_codes" property as channel count. 

It comes down to this:
Do I replace Kconfig options by device tree properties?

Reply:
Sorry down to the main question:

> Do I replace Kconfig options by device tree properties?

Yes I think it'd be a better solution, if anything because it means that the CI test build is always going to cover all the code. It's a more convoluted change but if you feel like doing I think it'd lead to a better driver. Otherwise if you can get that build assert sorted I think that'd be good enough. Your choice.